aips/aip-119.md

```diff
@@ -0,0 +1,230 @@
+---
+aip: (this is determined by the AIP Manager, leave it empty when drafting)
+title: Scheduled Transactions
+author: Manu Dhundi (https://github.com/manudhundi) & Zekun Li (https://github.com/zekun000)
+discussions-to (*optional):
+Status: Draft
+last-call-end-date (*optional):
+type: Core, Framework
+created: 04/15/2025
+updated (*optional):
+requires (*optional): <AIP number(s)>
+---
+
+# AIP-119 - Scheduled Transactions
+  
+(Please give a temporary file name to your AIP when first drafting it, such as `aip-x.md`. The AIP manager will assign a number to it after reviewing.)
+
+(Please remove the questions in the "quote box". Provide complete context to the questions being asked in the content that you provide.)
+
+## Summary
+A scheduled transaction is a user-submitted transaction that specifies another transaction to be executed at a future time. The system executes it in a suitable future block.  
+ 
+This AIP proposes framework changes to support scheduling, gas handling, and deposits for scheduling future transactions. It also introduces core changes for the blockchain to check at every block for ready scheduled transactions, prioritize them alongside others based on gas fees and existing rules, and execute them accordingly.
+
+### Out of scope
+ * Privacy
+ * Event based triggers
+ * Failure Retries
+ * Future execution gas markets
+ * Deterministic slicing
+
+## High-level Overview
+Framework provides a mechanism that enables users to schedule transactions for future execution programmatically. Users can submit a scheduling transaction specifying a target function, a future execution time (in milliseconds), the maximum gas they’re willing to spend, and the price per gas unit. An optional rescheduling delta can be included to enable recurring scheduling. To prevent denial-of-service or spam, users must deposit the full gas budget up front, which is held in a dedicated fungible asset store. Users may cancel scheduled transactions at any time before execution to reclaim their deposit.
+
+Once submitted, scheduled transactions are stored in a queue data structure sorted by execution time, with 100ms granularity (the order of expected block intervals). At the beginning of each block, the blockchain core system checks whether any scheduled transactions are due. If so, they are considered as candidate transactions and prioritized using specified price per gas unit and existing rules like say, [use case aware block reordering](#https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-68.md). Transactions that are not selected for execution are retried in future blocks, up to a fixed expiration window (around 100 blocks). 
```

Comment:
nit: should we consider all or nothing scheduling e.g., schedule in 100ms at the very next block, or cancel (instead of retrying up to some time/block window)?